src/main/java/de/tum/cit/aet/artemis/core/domain/LLMTokenUsage.java

```diff
@@ -0,0 +1,154 @@
+package de.tum.cit.aet.artemis.core.domain;
+
+import java.time.ZonedDateTime;
+
+import jakarta.annotation.Nullable;
+import jakarta.persistence.Column;
+import jakarta.persistence.Entity;
+import jakarta.persistence.EnumType;
+import jakarta.persistence.Enumerated;
+import jakarta.persistence.Inheritance;
+import jakarta.persistence.InheritanceType;
+import jakarta.persistence.JoinColumn;
+import jakarta.persistence.ManyToOne;
+import jakarta.persistence.Table;
+
+import org.hibernate.annotations.Cache;
+import org.hibernate.annotations.CacheConcurrencyStrategy;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+
+import de.tum.cit.aet.artemis.exercise.domain.Exercise;
+import de.tum.cit.aet.artemis.iris.domain.message.IrisMessage;
+
+@Entity
+@Table(name = "llm_token_usage")
+@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
+@Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+public class LLMTokenUsage extends DomainObject {
+
+    @Column(name = "service")
+    @Enumerated(EnumType.STRING)
+    private LLMServiceType serviceType;
+
+    @Column(name = "model")
+    private String model;
+
+    @Column(name = "cost_per_token")
+    private double cost_per_token;
+
+    @Column(name = "num_input_tokens")
+    private int num_input_tokens;
+
+    @Column(name = "num_output_tokens")
+    private int num_output_tokens;
+
+    @Nullable
+    @ManyToOne
+    @JsonIgnore
+    @JoinColumn(name = "course_id")
+    private Course course;
+
+    @Nullable
+    @ManyToOne
+    @JsonIgnore
+    @JoinColumn(name = "exercise_id")
+    private Exercise exercise;
+
+    @Column(name = "user_id")
+    private long userId;
+
+    @Nullable
+    @Column(name = "timestamp")
+    private ZonedDateTime timestamp = ZonedDateTime.now();
+
+    @Nullable
+    @ManyToOne
+    @JsonIgnore
+    @JoinColumn(name = "iris_message_id")
+    IrisMessage irisMessage;
```

Comment:
Possibly, a `trace_id` would be useful so that we can group multiple LLM calls from a single request. This way we can say, for example: "X requests were made, resulting in an average cost of Y per request."  `iris_message_id` kind of achieves this for Iris, but I think a `trace_id` would be beneficial for both Iris and Athena.

Reply:
I don't think that is necessary right now, I'd add it on top. 

But it would be good to have a helper function where you can store multiple LLM calls with the same `trace_id` at once.

Reply:
This is the current function that saves the usage:
```java
public List<LLMTokenUsage> saveTokenUsage(IrisMessage message, Exercise exercise, User user, Course course, List<PyrisLLMCostDTO> tokens) {
        List<LLMTokenUsage> tokenUsages = new ArrayList<>();
        for (PyrisLLMCostDTO cost : tokens) {
            LLMTokenUsage llmTokenUsage = new LLMTokenUsage();
            if (message != null) {
                llmTokenUsage.setIrisMessage(message);
                llmTokenUsage.setTimestamp(message.getSentAt());
            }
            llmTokenUsage.setServiceType(cost.pipeline());
            llmTokenUsage.setExercise(exercise);
            if (user != null) {
                llmTokenUsage.setUserId(user.getId());
            }
            llmTokenUsage.setCourse(course);
            llmTokenUsage.setNum_input_tokens(cost.num_input_tokens());
            llmTokenUsage.setNum_output_tokens(cost.num_output_tokens());
            llmTokenUsage.setModel(cost.model_info());
            tokenUsages.add(llmTokenUsageRepository.save(llmTokenUsage));
        }
        return tokenUsages;
    }
```
With IrisMessage, it is currently possible to group Iris calls using the same message. With the trace_id, I am unsure how to add the option to group them. Do I add another function where the trace_id can be set manually? Or what was your idea with the helper function?

I am trying to figure out how to track the trace_id for one call and have a unique trace_id for each call without another table with a primary key.

Reply:
I think you'd have to persist/save the first `LLMTokenUsage` and let the database generate an id for the `trace_id`, then you'd use this id for the remaining entries

Reply:
I haven't added a helper function, as the saveTokenUsage accepts a list where all the call tokens can be stored. What I added is a `Long traceId`, which is calculated before the iteration over the tokens and is unique as it is calculated with a random UUID and the current timestamp:
```java
        // Combine current time and UUID to create a unique traceId
        long timestamp = System.currentTimeMillis();
        long uuidComponent = UUID.randomUUID().getLeastSignificantBits() & Long.MAX_VALUE;
        Long traceId = timestamp + uuidComponent;
```
 That is then used for all the tokens that need to be stored to let them have the same traceId. Is that how you imagined it?